File: companion-apps/android/app/src/main/java/me/fiksu/esp32map/companion/domain/SharedImportModels.kt
package me.fiksu.esp32map.companion.domain

data class SharedImportEnvelope(
    val id: String,
    val sourceApplication: String?,
    val receivedAtEpochMs: Long,
    val rawKind: SharedImportRawKind,
    val mimeType: String?,
    val uniformTypeIdentifier: String?,
    val fileName: String?,
    val fileSizeBytes: Int?,
    val originalText: String?,
    val originalUrl: String?,
    val storedFilePath: String?,
    val classification: SharedImportClassification,
    val disposition: SharedImportDisposition,
    val note: String?,
)

enum class SharedImportRawKind {
    URL,
    PLAIN_TEXT,
    FILE,
    MULTIPLE_ITEMS,
    UNKNOWN,
}

enum class SharedImportClassification {
    GOOGLE_MAPS_LOCATION_LINK,
    GENERIC_LOCATION_LINK,
    GPX_FILE,
    GENERIC_XML_FILE,
    FIT_FILE,
    TCX_FILE,
    PLAIN_COORDINATES,
    UNSUPPORTED_UNKNOWN,
}

enum class SharedImportDisposition {
    DIRECT_HOME_PREVIEW,
    ROUTE_DETAIL_REVIEW,
    DIAGNOSTICS_ONLY,
}

data class ImportDiagnosticsEntry(
    val id: String,
    val envelope: SharedImportEnvelope,
    val createdAtEpochMs: Long,
) {
    val title: String
        get() = envelope.fileName
            ?: envelope.originalUrl?.let { runCatching { java.net.URI(it).host }.getOrNull() }
            ?: envelope.sourceApplication
            ?: envelope.classification.name

    val subtitle: String
        get() = envelope.note
            ?: envelope.originalText?.trim()?.takeIf { it.isNotEmpty() }?.take(120)
            ?: envelope.originalUrl?.take(120)
            ?: "Unsupported shared item"
}

File: companion-apps/ios/CompanionSharedImport/SharedImportModels.swift
import Foundation

enum SharedImportRawKind: String, Codable, Equatable {
    case url
    case plainText
    case file
    case multipleItems
    case unknown
}

enum SharedImportClassification: String, Codable, Equatable {
    case googleMapsLocationLink
    case genericLocationLink
    case gpxFile
    case genericXMLFile
    case fitFile
    case tcxFile
    case plainCoordinates
    case unsupportedUnknown
}

enum SharedImportDisposition: String, Codable, Equatable {
    case directHomePreview
    case routeDetailReview
    case diagnosticsOnly
}

struct SharedImportDebugContext: Codable, Equatable {
    var producerTarget: String?
    var producerBundleID: String?
    var producerVersion: String?
    var producerBuild: String?
    var latestHandlerTarget: String?
    var latestHandlerBundleID: String?
    var latestHandlerVersion: String?
    var latestHandlerBuild: String?
    var latestPhase: String?
    var latestOutcome: String?
    var lastUpdatedAt: Date?
}

struct SharedImportEnvelope: Identifiable, Codable, Equatable {
    var id: String
    var sourceApplication: String?
    var receivedAt: Date
    var rawKind: SharedImportRawKind
    var mimeType: String?
    var uniformTypeIdentifier: String?
    var fileName: String?
    var fileSizeBytes: Int?
    var originalText: String?
    var originalURL: String?
    var storedFilePath: String?
    var classification: SharedImportClassification
    var disposition: SharedImportDisposition
    var note: String?
    var debugContext: SharedImportDebugContext?
    var debugTrail: [String]?
}

struct ImportDiagnosticsEntry: Identifiable, Codable, Equatable {
    var id: String
    var envelope: SharedImportEnvelope
    var createdAt: Date

    var title: String {
        if let fileName = envelope.fileName, !fileName.isEmpty {
            return fileName
        }
        if let url = envelope.originalURL, let host = URL(string: url)?.host, !host.isEmpty {
            return host
        }
        if let sourceApplication = envelope.sourceApplication, !sourceApplication.isEmpty {
            return sourceApplication
        }
        return envelope.classification.rawValue
    }

    var subtitle: String {
        if let note = envelope.note, !note.isEmpty {
            return note
        }
        if let text = envelope.originalText?.trimmingCharacters(in: .whitespacesAndNewlines), !text.isEmpty {
            return String(text.prefix(120))
        }
        if let url = envelope.originalURL, !url.isEmpty {
            return String(url.prefix(120))
        }
        return "Unsupported shared item"
    }

    var debugPackageText: String {
        struct DebugPackage: Codable {
            var formatVersion: Int
            var exportedAt: Date
            var entry: ImportDiagnosticsEntry
        }
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let package = DebugPackage(formatVersion: 2, exportedAt: Date(), entry: self)
        let json = (try? encoder.encode(package)).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
        return json
    }
}
